src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,132 @@
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            // initialize public properties
+            _sortOption = SortOption.Location;
+            _documentSymbolUIItems = new ObservableCollection<DocumentSymbolUIItem>();
+
+            // setup work queues
+            _documentSymbolQueue = new AsyncBatchingWorkQueue<VisualStudioCodeWindowInfo, DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                GetDocumentSymbolAsync,
+                EqualityComparer<VisualStudioCodeWindowInfo>.Default,
+                asyncListener,
+                CancellationToken);
+
+            _filterAndSortQueue = new AsyncBatchingWorkQueue<FilterAndSortOptions, DocumentSymbolDataModel?>(
+                DelayTimeSpan.NearImmediate,
+                FilterAndSortDataModelAsync,
+                EqualityComparer<FilterAndSortOptions>.Default,
+                asyncListener,
+                CancellationToken);
+
+            _updateUIQueue = new AsyncBatchingWorkQueue<UIData>(
+                DelayTimeSpan.NearImmediate,
+                UpdateUIAsync,
+                asyncListener,
+                CancellationToken);
+
+            _textViewEventSource.Changed += OnEventSourceChanged;
+            _textViewEventSource.Connect();
+
+            // queue initial model update
+            EnqueueModelUpdateAsync().Forget();
+        }
+
+        private string? _searchText;
+        public string? SearchText
+        {
+            get
+            {
+                return _searchText;
+            }
+            set
+            {
+                if (_searchText != value)
+                {
+                    _searchText = value;
+                    NotifyPropertyChanged();
+                }
+            }
+        }
+
+        private SortOption _sortOption;
+        public SortOption SortOption
+        {
+            get
+            {
+                return _sortOption;
+            }
+            set
+            {
+                if (_sortOption != value)
+                {
+                    _sortOption = value;
+                    NotifyPropertyChanged();
+                }
+            }
+        }
+
+        private ObservableCollection<DocumentSymbolUIItem> _documentSymbolUIItems;
+        public ObservableCollection<DocumentSymbolUIItem> DocumentSymbolUIItems
+        {
+            get => _documentSymbolUIItems;
+            set
+            {
+                _documentSymbolUIItems = value;
+                NotifyPropertyChanged();
+            }
+        }
+
+        private void OnEventSourceChanged(object sender, TaggerEventArgs e)
+            => EnqueueModelUpdateAsync().Forget();
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        private void NotifyPropertyChanged([CallerMemberName] string? propertyName = null)
+            => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+
+        public void Dispose()
+        {
```

Comment:
threading asserts.

Reply:
if that's the case, then we need to make the impl threadsafe.  One way is making everything readonly.  But as this is a mutable type, we likely will need locks to ensure proper state updates of this free threaded object.